src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -151,6 +168,249 @@ void ReadFromRemote::enforceAggregationInOrder()
+                /// SELECT x, y FROM (
+                ///     SELECT number + 1 AS x, sum(number) AS y FROM remote('127.0.0.{1,2}', numbers(3)) GROUP BY x
+                /// ) WHERE x = 1
+                /// In this case, ReadFromRemote has `plus(__table1.number, 1_UInt8) UInt64` in header,
+                /// and filter DAG has input column with name `plus(__table1.number, 1_UInt8)`.
+                /// Here, filter is pushed down before the aggregation, and projection name can't be used.
+                /// However, we can match the input with the execution name of projection query tree.
+                /// Note: this may not cover all the cases.
+                auto it = execution_name_to_projection_query_tree.find(node->result_name);
+                if (it != execution_name_to_projection_query_tree.end())
+                    /// Append full expression as an AST.
+                    /// We rely on plan optimization that the result is (expected to be) valid.
+                    res = it->second->toAST();
+            }
+        }
+
+        if (node->type == ActionsDAG::ActionType::ALIAS)
+            res = node_to_ast[node->children.at(0)];
+
+        if (node->type != ActionsDAG::ActionType::FUNCTION)
+            continue;
+
+        if (!node->function_base->isDeterministic() || node->function_base->isStateful())
+            continue;
+
+        bool has_all_args = true;
+        ASTs arguments;
+        for (const auto * child : node->children)
+        {
+            auto ast = node_to_ast[child];
+            if (!ast)
+                has_all_args = false;
+            else
+                arguments.push_back(std::move(ast));
+        }
+
+        /// Allow to skip children only for AND function.
+        auto func_name = node->function_base->getName();
+        bool is_function_and = func_name == "and";
+        if (!has_all_args && !is_function_and)
+            continue;
+
+        if (is_function_and && arguments.empty())
+            continue;
+
+        /// and() with 1 arg is not allowed. Make it AND(condition, 1)
+        if (is_function_and && arguments.size() == 1)
+            arguments.push_back(std::make_shared<ASTLiteral>(Field(1)));
+
+        /// Support for GLOBAL IN.
+        if (external_tables && isNameOfGlobalInFunction(func_name))
+        {
+            const auto * second_arg = node->children.at(1);
+            auto maybe_set = second_arg->column;
+            if (const auto * col_const = typeid_cast<const ColumnConst *>(maybe_set.get()))
+                maybe_set = col_const->getDataColumnPtr();
+
+            if (const auto * col_set = typeid_cast<const ColumnSet *>(maybe_set.get()))
+            {
+                auto future_set = col_set->getData();
+                if (auto * set_from_subquery = typeid_cast<FutureSetFromSubquery *>(future_set.get()))
+                {
+                    const auto temporary_table_name = fmt::format("_data_{}", toString(set_from_subquery->getHash()));
```

Comment:
Why don't you use  `FutureSetFromSubquery::external_table` here?

Reply:
I guess it does not matter ...